lib/reversetunnel/track/tracker.go

```diff
@@ -0,0 +1,303 @@
+	for key, set := range p.sets {
+		if set.expire(cutoff) > 0 {
+			count := len(set.proxies)
+			if count < 1 {
+				count = 1
+			}
+			p.wp.Set(key, uint64(count))
+		}
+	}
+}
+
+func (p *Tracker) getOrCreate(key Key) *proxySet {
+	if s, ok := p.sets[key]; ok {
+		return s
+	}
+	set := newProxySet(key)
+	p.sets[key] = set
+	p.wp.Set(key, 1)
+	return set
+}
+
+// WithProxy runs the supplied closure if and only if
+// no other work is currently being done with the proxy
+// identified by principals.
+func (p *Tracker) WithProxy(work func(), lease Lease, principals ...string) (didWork bool) {
+	key := lease.Key().(Key)
+	if ok := p.claim(key, principals...); !ok {
+		return false
+	}
+	defer p.unclaim(key, principals...)
+	work()
+	return true
+}
+
+func (p *Tracker) claim(key Key, principals ...string) (ok bool) {
+	p.Lock()
+	defer p.Unlock()
+	set, ok := p.sets[key]
+	if !ok {
+		return false
+	}
+	return set.claim(principals...)
+}
+
+func (p *Tracker) unclaim(key Key, principals ...string) {
+	p.Lock()
+	defer p.Unlock()
+	set, ok := p.sets[key]
+	if !ok {
+		return
+	}
+	set.unclaim(principals...)
+}
+
+type entry struct {
+	lastSeen time.Time
+	claimed  bool
+}
+
+func newProxySet(key Key) *proxySet {
+	return &proxySet{
+		key:     key,
+		proxies: make(map[string]entry),
+	}
+}
+
+type proxySet struct {
+	key     Key
+	proxies map[string]entry
+}
+
+func (p *proxySet) claim(principals ...string) (ok bool) {
+	proxy := p.resolveName(principals)
+	e, ok := p.proxies[proxy]
+	if !ok {
+		p.proxies[proxy] = entry{
+			claimed: true,
+		}
+		return true
+	}
+	if e.claimed {
+		return false
+	}
+	e.claimed = true
+	p.proxies[proxy] = e
+	return true
+}
+
+func (p *proxySet) unclaim(principals ...string) {
+	proxy := p.resolveName(principals)
+	p.proxies[proxy] = entry{
+		lastSeen: time.Now(),
+	}
+}
+
+func (p *proxySet) markSeen(t time.Time, proxy string) {
+	e, ok := p.proxies[proxy]
+	if !ok {
+		p.proxies[proxy] = entry{
+			lastSeen: t,
+		}
+		return
+	}
+	if e.lastSeen.After(t) {
+		return
+	}
+	e.lastSeen = t
+	p.proxies[proxy] = e
+}
+
+func (p *proxySet) expire(cutoff time.Time) (removed int) {
+	for name, entry := range p.proxies {
+		if entry.claimed {
+			continue
+		}
+		if entry.lastSeen.Before(cutoff) {
+			delete(p.proxies, name)
+			removed++
+		}
+	}
+	return
+}
+
+// resolveName tries to extract the UUID of the proxy as that's the
+// only unique identifier in the list of principals.
+func (p *proxySet) resolveName(principals []string) string {
+	// check if we're already using one of these principals.
+	for _, name := range principals {
+		if _, ok := p.proxies[name]; ok {
+			return name
+		}
+	}
+	// default to using the first principal
+	name := principals[0]
+	// if we have a `.<cluster-name>` suffix, remove it.
+	if strings.HasSuffix(name, p.key.Cluster) {
+		t := strings.TrimSuffix(name, p.key.Cluster)
+		if strings.HasSuffix(t, ".") {
+			name = strings.TrimSuffix(t, ".")
+		}
+	}
```

Comment:
How about:
```go
name = strings.TrimSuffix(name, fmt.Sprintf(".%s", p.key.Cluster))
```
it will be a noop if there's no such suffix

Reply:
Trimming `Cluster` and `"."` separately avoids unnecessary string allocation.